src/ram/transform/MakeIndex.cpp

```diff
@@ -38,138 +38,235 @@
+            const RamExpression* lhs = &binRelOp->getLHS();
+            if (rhs->getTupleId() == identifier && rla->getLevel(lhs) < identifier) {
+                element = rhs->getElement();
+                return {std::make_unique<RamUndefValue>(), clone(lhs)};
+            }
+        }
+    }
+    return {std::make_unique<RamUndefValue>(), std::make_unique<RamUndefValue>()};
+}
 
-        if (isGreaterThanConstraint(binRelOp->getOperator())) {
-            // Tuple[level, element] > <expr>
-            // is equivalent to
-            // Tuple[level, element] >= <expr> + 1
+// Retrieves the <expr1> <= Tuple[level, element] <= <expr2> part of the constraint as a pair { <expr1>,
+// <expr2> }
+ExpressionPair MakeIndexTransformer::getLowerUpperExpression(
+        RamCondition* c, size_t& element, int identifier) {
+    if (auto* binRelOp = dynamic_cast<RamConstraint*>(c)) {
+        // TODO: FIXME: how does this interact w/ `FEQ`?
+
+        if (isEqConstraint(binRelOp->getOperator())) {
             if (const auto* lhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getLHS())) {
                 const RamExpression* rhs = &binRelOp->getRHS();
                 if (lhs->getTupleId() == identifier && rla->getLevel(rhs) < identifier) {
                     element = lhs->getElement();
-                    std::vector<std::unique_ptr<RamExpression>> expressions;
-                    expressions.push_back(souffle::clone(rhs));
-                    expressions.push_back(std::make_unique<RamSignedConstant>(RamDomain(1)));
-
-                    return {std::make_unique<RamIntrinsicOperator>(FunctorOp::ADD, std::move(expressions)),
-                            std::make_unique<RamUndefValue>()};
+                    return {clone(rhs), clone(rhs)};
                 }
             }
-            // <expr> > Tuple[level, element]
-            // is equivalent to
-            // <expr> - 1 >= Tuple[level, element]
             if (const auto* rhs = dynamic_cast<const RamTupleElement*>(&binRelOp->getRHS())) {
                 const RamExpression* lhs = &binRelOp->getLHS();
                 if (rhs->getTupleId() == identifier && rla->getLevel(lhs) < identifier) {
                     element = rhs->getElement();
-                    std::vector<std::unique_ptr<RamExpression>> expressions;
-                    expressions.push_back(souffle::clone(lhs));
-                    expressions.push_back(std::make_unique<RamSignedConstant>(RamDomain(1)));
-
-                    return {std::make_unique<RamUndefValue>(),
-                            std::make_unique<RamIntrinsicOperator>(FunctorOp::SUB, std::move(expressions))};
+                    return {clone(lhs), clone(lhs)};
                 }
             }
+        } else if (isIneqSigned(binRelOp->getOperator())) {
+            return getSignedExpressionPair(binRelOp, element, identifier);
         }
+        /* TODO: Fix Comparators in BTree to handle types
```

Comment:
Tag the todo with a username or issue number. Do not leave in the commented out code.